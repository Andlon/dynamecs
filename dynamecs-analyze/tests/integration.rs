use escargot::CargoBuild;
use insta::assert_snapshot;
use tempfile::tempdir;
use time::{Date, Month, UtcOffset};
use dynamecs_analyze::{iterate_records, iterate_records_from_reader, Record, RecordBuilder, write_records};

fn replace_middle(s: &str, prefix: &str, suffix: &str, replacement: &str) -> Option<String> {
    s.strip_prefix(prefix)
        .and_then(|s| s.strip_suffix(suffix))
        .map(|_| format!("{prefix}{replacement}{suffix}"))
}

fn redact_message(msg: &str) -> String {
    // Replace the "middle" part of the pattern "<prefix><middle><suffix>"
    let path_replacement = "<redacted path>";
    let path_redactions = [
        // prefix, suffix
        ["Working directory: ", ""],
        ["Logging text to file ", " with log level trace"],
        ["Logging JSON to file ", " with log level trace"],
        ["Archived log file path: ", ""],
        ["Archived JSON log file path: ", ""],
        ["Output base path: ", ""],
    ];
    for [prefix, suffix] in path_redactions {
        if let Some(replaced) = replace_middle(msg, prefix, suffix, path_replacement) {
            return replaced;
        }
    }

    msg.to_string()
}

fn redact_records(records: &[Record]) -> Vec<Record> {
    let arbitrary_timestamp = Date::from_calendar_date(2000, Month::November, 14)
        .unwrap()
        .with_hms(08, 00, 00)
        .unwrap()
        .assume_offset(UtcOffset::from_hms(02, 00, 00).unwrap());
    records.iter()
        .cloned()
        .map(|record| {
            let message_override = record.message()
                .map(|message| redact_message(message));

            let mut builder = RecordBuilder::from_record(record)
                .timestamp(arbitrary_timestamp)
                .thread_id("ThreadId(0)");

            if let Some(msg) = message_override {
                builder = builder.message(msg);
            }

            builder.build()
        })
        .collect()
}

#[test]
fn test_expected_output_for_basic_app1() -> Result<(), Box<dyn std::error::Error>> {
    // We test that the JSON records that we manually write with write_records
    // are equivalent when imported again to the expected output from an artificial,
    // minimal dynamecs-app application
    let temp_dir = tempdir()?;
    let target_dir = temp_dir.path().join("target");
    let output_dir = temp_dir.path().join("output");
    let _ = CargoBuild::new()
        .manifest_path("../test-apps/Cargo.toml")
        .bin("basic_app1")
        .target_dir(target_dir)
        .run()?
        .command()
        .args(["--output-dir", output_dir.to_str().unwrap()])
        .args(["--file-log-level", "trace"])
        .output()?;

    let records: Vec<_> = iterate_records(output_dir.join("logs/dynamecs_app.jsonlog"))?
        .map(|record_result| record_result.expect("The records are complete/correct for this test"))
        .collect();
    let redacted_records = redact_records(&records);

    let mut bytes: Vec<u8> = vec![];
    write_records(&mut bytes, redacted_records.clone().into_iter())?;
    let records_string = String::from_utf8(bytes.clone())?;

    assert_snapshot!(records_string);

    let records_roundtrip: Vec<_> = iterate_records_from_reader(bytes.as_slice())
        .map(|record_result| record_result.expect("Records must be complete"))
        .collect();

    assert_eq!(records_roundtrip, redacted_records);

    Ok(())
}