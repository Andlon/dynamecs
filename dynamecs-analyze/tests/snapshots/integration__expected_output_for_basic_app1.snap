---
source: dynamecs-analyze/tests/integration.rs
expression: records_string
---
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Working directory: <redacted path>"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Logging text to stdout with log level info"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Logging text to file <redacted path> with log level trace"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Logging JSON to file <redacted path> with log level trace"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Archived log file path: <redacted path>"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Archived JSON log file path: <redacted path>"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Output base path: <redacted path>"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"No configuration specified. Trying to use the empty document {} as default."},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Using configuration: \n{}"},"target":"dynamecs_app","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Initializing scenario"},"target":"basic_app1","threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Starting simulation of scenario \"basic_app1\""},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"step","step_index":0},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"DEBUG","fields":{"message":"Running post-systems for initial state"},"target":"dynamecs_app","span":{"name":"step","step_index":0},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"post_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"post_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"post_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Starting step 0 at simulation time 0.00000 (dt = 1.00000e-1)"},"target":"dynamecs_app","span":{"name":"step","step_index":0},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"pre_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"pre_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"pre_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"simulation_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"DEBUG","fields":{"answer":42,"message":"debug-test"},"target":"target1","span":{"name":"simulation_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"target1","span":{"name":"span1"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"},{"name":"span1"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"target2","span":{"name":"span2"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"},{"name":"span1"},{"name":"span2"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"TRACE","fields":{"message":"trace-test","question":"jeopardy"},"target":"target2","span":{"name":"span2"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"},{"name":"span1"},{"name":"span2"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"target2","span":{"name":"span2"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"},{"name":"span1"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"target1","span":{"name":"span1"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"simulation_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"simulation_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"post_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0},{"name":"post_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"post_systems"},"spans":[{"name":"run"},{"name":"step","step_index":0}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"step","step_index":0},"spans":[{"name":"run"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"step","step_index":1},"spans":[{"name":"run"},{"name":"step","step_index":1}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Starting step 1 at simulation time 0.10000 (dt = 1.00000e-1)"},"target":"dynamecs_app","span":{"name":"step","step_index":1},"spans":[{"name":"run"},{"name":"step","step_index":1}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"pre_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"pre_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"pre_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"simulation_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"DEBUG","fields":{"answer":42,"message":"debug-test"},"target":"target1","span":{"name":"simulation_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"target1","span":{"name":"span1"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"},{"name":"span1"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"target2","span":{"name":"span2"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"},{"name":"span1"},{"name":"span2"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"TRACE","fields":{"message":"trace-test","question":"jeopardy"},"target":"target2","span":{"name":"span2"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"},{"name":"span1"},{"name":"span2"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"target2","span":{"name":"span2"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"},{"name":"span1"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"target1","span":{"name":"span1"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"simulation_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"simulation_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"post_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1},{"name":"post_systems"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"post_systems"},"spans":[{"name":"run"},{"name":"step","step_index":1}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"step","step_index":1},"spans":[{"name":"run"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"Simulation ended"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}],"threadId":"ThreadId(0)"}
{"timestamp":"2000-11-14T08:00:00+02:00","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"run"},"spans":[],"threadId":"ThreadId(0)"}

